transformer_engine/pytorch/attention.py

```diff
@@ -515,22 +516,41 @@ class AttnFuncWithCP(torch.autograd.Function):
 
     @staticmethod
     def forward(ctx, is_training, q, k, v, cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k,
-                dropout_p, cp_group, cp_global_ranks, cp_stream, softmax_scale, attn_mask_type,
-                deterministic, use_fused_attention):
+                dropout_p, cp_group, cp_global_ranks, cp_stream, softmax_scale, qkv_format,
+                attn_mask_type, attn_bias_type, attn_bias, deterministic, use_fused_attention):
         if softmax_scale is None:
             softmax_scale = q.shape[-1] ** (-0.5)
 
         cp_size = get_distributed_world_size(cp_group)
         rank = get_distributed_rank(cp_group)
         send_dst = cp_global_ranks[(rank + 1) % cp_size]
-        recv_src = cp_global_ranks[(rank + cp_size - 1) % cp_size]
+        recv_src = cp_global_ranks[(rank - 1) % cp_size]
         batch_p2p_comm = int(os.getenv("NVTE_BATCH_MHA_P2P_COMM", "0")) or (cp_size == 2)
 
         causal = (attn_mask_type == "causal")
 
+        qkv_layout = qkv_format + "_" + qkv_format + "_" + qkv_format
+
         if causal:
-            # [b, s, np, hn] -> [b, 2, s//2, np, hn]
-            q, k, v = [x.view(x.shape[0], 2, x.shape[1]//2, *x.shape[2:]) for x in [q, k, v]]
+            if qkv_format == "bshd":
+                # [b, s, np, hn] -> [b, 2, s//2, np, hn]
+                q, k, v = [x.view(x.shape[0], 2, x.shape[1]//2, *x.shape[2:]) for x in [q, k, v]]
+            elif qkv_format == "sbhd":
+                # [s, b, np, hn] -> [2, s//2, b, np, hn]
+                q, k, v = [x.view(2, x.shape[0]//2, *x.shape[1:]) for x in [q, k, v]]
+        if attn_bias is not None:
+            assert (len(attn_bias.shape) == 4), "Attention bias shape should be [b, h, sq, sk]"
```

Comment:
cuDNN supports forward for `[b, h, sq, sk]` but only `[1, h, sq, sk]` for backward. Maybe change the error message here a bit? 

https://github.com/xrennvidia/TransformerEngine/blob/85c10dd377d4849c4300101cd4082d877c4aad36/transformer_engine/pytorch/attention.py#L3707

Reply:
Done